### DISCUSSION TITLE
Configuring HTTP POST authorization

### DISCUSSION THREAD

**Discussion by agronholm** [emqx/emqx#9127] [Q&A] [answered]
I've been trying to configure a POST based HTTP authorization hook (got authentication to work already) with the following configuration:
```
    EMQX_LOADED_PLUGINS: emqx_auth_http,emqx_dashboard
    EMQX_CLUSTER__DISCOVERY_STRATEGY: manual
    EMQX_LISTENERS__SSL__DEFAULT__SSL_OPTIONS__CERTFILE: /tmp/ssl/tls.crt
    EMQX_LISTENERS__SSL__DEFAULT__SSL_OPTIONS__KEYFILE: /tmp/ssl/tls.key
    EMQX_LISTENERS__SSL__DEFAULT__SSL_OPTIONS__VERSIONS: "[tlsv1.3, tlsv1.2]"
    EMQX_LISTENERS__WSS__DEFAULT__SSL_OPTIONS__CERTFILE: /tmp/ssl/tls.crt
    EMQX_LISTENERS__WSS__DEFAULT__SSL_OPTIONS__KEYFILE: /tmp/ssl/tls.key
    EMQX_LISTENERS__WSS__DEFAULT__SSL_OPTIONS__VERSIONS: "[tlsv1.3, tlsv1.2]"
    EMQX_AUTHENTICATION__1__MECHANISM: password_based
    EMQX_AUTHENTICATION__1__BACKEND: http
    EMQX_AUTHENTICATION__1__METHOD: post
    EMQX_AUTHENTICATION__1__ENABLE: "true"
    EMQX_AUTHENTICATION__1__URL: http://web:8800/v1/integrations/mqtt/authenticate
    EMQX_AUTHENTICATION__1__BODY: >-
      {\"username\": \"${username}\",\"password\": \"${password}\", \"cert_subject\": \"\", \"peerhost\": \"${peerhost}\"}
    EMQX_AUTHORIZATION__NO_MATCH: deny
    EMQX_AUTHORIZATION__CACHE__MAX_SIZE: 1024
    EMQX_AUTHORIZATION__SOURCES__1__ENABLE: "false"
    EMQX_AUTHORIZATION__SOURCES__2__TYPE: http
    EMQX_AUTHORIZATION__SOURCES__2__METHOD: post
    EMQX_AUTHORIZATION__SOURCES__2__ENABLE: "true"
    EMQX_AUTHORIZATION__SOURCES__2__URL: http://web:8800/v1/integrations/mqtt/authorize
    EMQX_AUTHORIZATION__SOURCES__2__BODY: >-
      {\"username\": \"${username}\",\"password\": \"${password}\", \"cert_subject\": \"${cert_subject}\", \"topic\": \"${topic}\", \"action\": \"${action}\"}
```
Trouble is, I get a `GET` type authorization config instead:
![image](https://user-images.githubusercontent.com/130003/194883976-a2e61f52-b450-4a3c-b490-11e28fb59428.png)
Am I doing something wrong? This is against EMQX 5.0.8.

**Comment by agronholm** [comment]
Hm, I deleted the authorizer and, did a `helm upgrade` and then deleted the pod, but the http authorizer did not reappear. So it seems that the configuration is only applied on the first run.

**Reply by agronholm** [reply]
I think this question becomes: how do I go about regenerating the config based on the environment variables?

**Reply by id** [reply]
@agronholm could you try with the latest version?

**Reply by agronholm** [reply]
Yeah, this was actually resolved some time ago. I noticed when my config stopped working, and I needed to use the authorization source at index 1 instead of 2. But we no longer use environment variables for this anyway, since the hooks were problematic in a CI environment and we had to give up on them.

**Comment by vec715** [comment]
is there any updates on this? have the same issue, trying to configure HTTP POST Body but getting the empty form

**Reply by id** [reply]
@vec715 could you describe your setup, version, and config?

**Comment by zmstone** [comment, marked as answer]
I don't think you need this:
`EMQX_AUTHORIZATION__SOURCES__1__ENABLE: "false"`
The 1st authz source is an implicit default in the software, but not some pre-configured default in a config file,
so you can just go ahead setting http post as `__1__`

**Comment by zmstone** [comment]
> Hm, I deleted the authorizer and, did a helm upgrade and then deleted the pod, but the http authorizer did not reappear. So it seems that the configuration is only applied on the first run.

For this question. I tried to reproduce it with `emqx/emqx:5.0.8` image with below docker run:

```
docker run --rm -it -v $(pwd)/t:/opt/emqx/data --name emqx --net host \
-e EMQX_AUTHORIZATION__SOURCES__1__ENABLE="false" \
-e EMQX_AUTHORIZATION__SOURCES__2__TYPE=http \
-e EMQX_AUTHORIZATION__SOURCES__2__METHOD=post \
-e EMQX_AUTHORIZATION__SOURCES__2__ENABLE="true" \
-e EMQX_AUTHORIZATION__SOURCES__2__URL="http://web:8800/v1/integrations/mqtt/authorize" \
-e EMQX_AUTHORIZATION__SOURCES__2__BODY='{}' \
emqx/emqx:5.0.8
```

1. Start docker
2. Delete the "file" authorizer from dashbaord
3. Stop the docker and start again with the exact same command above
4. Inspect authorizers on dashboard

What I observed: there is only http authorizer which is in disabled state (because of the `EMQX_AUTHORIZATION__SOURCES__1__ENABLE: "false"` env variable).
`__2__` is not added because it only allows one authorizer of each type in the chain.
